satsense/features/texton.py

```diff
@@ -58,22 +59,34 @@ def get_texton_descriptors(image: Image):
 Image.register('texton_descriptors', get_texton_descriptors)
 
 
-def texton_cluster(images: Iterator[Image], n_clusters=32,
-                   sample_size=100000) -> MiniBatchKMeans:
+def texton_cluster(images: Iterator[Image],
+                   n_clusters=32,
+                   max_samples=100000,
+                   sample_window=(8192, 8192)) -> MiniBatchKMeans:
     """Compute texton clusters."""
+    nfeatures = int(max_samples / len(images))
     descriptors = []
     for image in images:
-        array = image['texton_descriptors']
-        array = array.reshape(-1, array.shape[-1])
-        non_masked = ~array.mask.any(axis=-1)
-        descriptors.append(array.data[non_masked])
-    descriptors = np.vstack(descriptors)
+        chunk = np.minimum(image.shape, sample_window)
+
+        generator = FullGenerator(image, chunk)
+        generator.load_image('texton_descriptors', (chunk, ))
+
+        max_features_per_window = int(nfeatures / np.prod(generator.shape))
 
-    if descriptors.shape[0] > sample_size:
-        # Limit the number of descriptors to sample_size
-        # by randomly selecting some rows
-        descriptors = descriptors[np.random.choice(
-            descriptors.shape[0], sample_size, replace=False), :]
+        img_descriptors = []
```

Comment:
I think `img_descriptors` is not needed, you can just directly append to `descriptors`